pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcResourceWriter.java

```diff
@@ -85,65 +97,92 @@ static void createTables(FhirEtlOptions options)
-        statement =
-            connection.prepareStatement(
-                "INSERT INTO "
-                    + tableName
-                    + " (id, type, datab) VALUES(?, ?, ?::jsonb) "
-                    + "ON CONFLICT (id, type) DO UPDATE SET id=?, type=?, datab=?::jsonb ;");
-      } else {
-        statement =
+  public void writeResource(Resource resource) throws SQLException, ViewApplicationException {
+    if (viewManager == null) {
+      try (Connection connection = jdbcDataSource.getConnection()) {
+        String tableName = resource.getResourceType().name();
+        PreparedStatement statement =
             connection.prepareStatement(
                 "INSERT INTO "
                     + tableName
                     + " (id, datab) VALUES(?, ?::jsonb) "
                     + "ON CONFLICT (id) DO UPDATE SET id=?, datab=?::jsonb ;");
+        statement.setString(1, resource.getIdElement().getIdPart());
+        statement.setString(2, parser.encodeResourceToString(resource));
+        statement.setString(3, resource.getIdElement().getIdPart());
+        statement.setString(4, parser.encodeResourceToString(resource));
+        statement.execute();
+        statement.close();
       }
-      int paramInd = 1;
-      statement.setString(paramInd, resource.getIdElement().getIdPart());
-      if (useSingleTable) {
-        paramInd++;
-        statement.setString(paramInd, resource.getResourceType().name());
-      }
-      paramInd++;
-      statement.setString(paramInd, parser.encodeResourceToString(resource));
-      paramInd++;
-      statement.setString(paramInd, resource.getIdElement().getIdPart());
-      if (useSingleTable) {
-        paramInd++;
-        statement.setString(paramInd, resource.getResourceType().name());
+    } else {
+      ImmutableList<ViewDefinition> views = viewManager.getViewsForType(resource.fhirType());
+      if (views != null) {
+        for (ViewDefinition vDef : views) {
+          if (Strings.isNullOrEmpty(vDef.getName())) {
+            throw new SQLException("Field `name` in ViewDefinition is not defined.");
+          }
+          ViewApplicator applicator = new ViewApplicator(vDef);
+          RowList rowList = applicator.apply(resource);
+          for (FlatRow row : rowList.getRows()) {
+            StringBuilder builder = new StringBuilder("INSERT INTO ");
+            builder.append(vDef.getName()).append(" (");
+            builder.append(String.join(",", rowList.getColumnNames()));
+            builder.append(") VALUES(");
+            // TODO add resource ID requirement and replacing old rows for incremental update.
+            builder.append(
+                String.join(
+                    ",", row.getElements().stream().map(e -> "?").collect(Collectors.toList())));
+            builder.append(");");
+            try (Connection connection = jdbcDataSource.getConnection()) {
+              PreparedStatement statement = connection.prepareStatement(builder.toString());
```

Comment:
Can we also attach the creation of `PreparedStatement` in the `try with resources` block, so that we can avoid the chances of not executing the `statement.close()` command (avoid leakages).

```
try (Connection connection = jdbcDataSource.getConnection(); PreparedStatement statement = connection.prepareStatement(builder.toString())) {
..
..
}
```

Reply:
Done.

Reply:
Thanks